File: go/parsers/unregister.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Unregister struct {
	gen *gen.Unregister
}

func NewUnregisterFields(g *gen.Unregister) messages.UnregisterFields {
	return &Unregister{gen: g}
}

func (u *Unregister) RequestID() uint64 {
	return u.gen.RequestID()
}

func (u *Unregister) RegistrationID() uint64 {
	return u.gen.RegistrationID()
}

func UnregisterToCapnproto(m *messages.Unregister) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	unregister, err := gen.NewRootUnregister(seg)
	if err != nil {
		return nil, err
	}

	unregister.SetRequestID(m.RequestID())
	unregister.SetRegistrationID(m.RegistrationID())

	data, err := msg.Marshal()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeUnregister, data), nil
}

func CapnprotoToUnregister(data []byte) (*messages.Unregister, error) {
	msg, err := capnp.Unmarshal(data)
	if err != nil {
		return nil, err
	}

	unregister, err := gen.ReadRootUnregister(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewUnregisterWithFields(NewUnregisterFields(&unregister)), nil
}

File: python/wamp_msgs_capnp/parsers/unregister.py
import os
from pathlib import Path

import capnp
from wampproto.messages.unregister import Unregister, IUnregisterFields

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
schema_file = os.path.join(root_dir, "schemas", "unregister.capnp")
unregister_capnp = capnp.load(str(schema_file))


class UnregisterFields(IUnregisterFields):
    def __init__(self, gen):
        self._gen = gen

    @property
    def request_id(self) -> int:
        return self._gen.requestID

    @property
    def registration_id(self) -> int:
        return self._gen.registrationID


def unregister_to_capnproto(u: Unregister) -> bytes:
    unregister = unregister_capnp.Unregister.new_message()
    unregister.requestID = u.request_id
    unregister.registrationID = u.registration_id

    packed_data = unregister.to_bytes_packed()
    return helpers.prepend_header(Unregister.TYPE, packed_data, b"")


def capnproto_to_unregister(data: bytes) -> Unregister:
    message_data, _ = helpers.extract_message(data)
    unregister_obj = unregister_capnp.Unregister.from_bytes_packed(message_data)

    return Unregister(UnregisterFields(unregister_obj))
